modAionImpl/src/org/aion/zero/impl/sync/TaskImportTrieData.java

```diff
@@ -0,0 +1,141 @@
+
+            // skip batch if everything already imported
+            if (nodes.isEmpty()) {
+                continue;
+            }
+
+            DatabaseType dbType = tnw.getDbType();
+            String peer = tnw.getDisplayId();
+            ByteArrayWrapper key;
+            byte[] value;
+            boolean failed = false;
+
+            for (Entry<ByteArrayWrapper, byte[]> e : nodes.entrySet()) {
+                key = e.getKey();
+                value = e.getValue();
+
+                TrieNodeResult result = chain.importTrieNode(key.getData(), value, dbType);
+
+                if (result.isSuccessful()) {
+                    fastSyncMgr.addImportedNode(key, value, dbType);
+                    log.debug(
+                            "<import-trie-nodes: key={}, value length={}, db={}, result={}, peer={}>",
+                            key,
+                            value.length,
+                            dbType,
+                            result,
+                            peer);
+                } else {
+                    if (log.isDebugEnabled()) {
+                        log.debug(
+                                "<import-trie-nodes-failed: key={}, value={}, db={}, result={}, peer={}>",
+                                key,
+                                Hex.toHexString(value),
+                                dbType,
+                                result,
+                                peer);
+                    }
+                    fastSyncMgr.handleFailedImport(key, value, dbType, tnw.getPeerId(), peer);
+                    failed = true;
+                    // exit this loop and ignore other imports
+                    break;
+                }
+            }
+
+            if (!failed) {
+                // reexamine missing states and make further requests
+                fastSyncMgr.updateRequests(
+                        tnw.getNodeKey(), tnw.getReferencedNodes().keySet(), tnw.getDbType());
+            }
+        }
+
+        if (log.isDebugEnabled()) {
+            log.debug("<import-trie-nodes: shutdown>");
+        }
+    }
+
+    /**
+     * Filters out trie nodes that have been imported when both the key and the value match the
+     * already imported data.
+     *
+     * @param wrapper the initial set of trie nodes to be imported
+     * @return the remaining nodes after the exact matches have been filtered out
+     */
+    private Map<ByteArrayWrapper, byte[]> filterImported(TrieNodeWrapper wrapper) {
+        Map<ByteArrayWrapper, byte[]> nodes =
+                wrapper.getReferencedNodes()
+                        .entrySet()
+                        .parallelStream()
+                        .filter(e -> !fastSyncMgr.containsExact(e.getKey(), e.getValue()))
+                        .collect(Collectors.toMap(Entry::getKey, Entry::getValue));
+
+        if (!fastSyncMgr.containsExact(wrapper.getNodeKey(), wrapper.getNodeValue())) {
```

Comment:
any reason for doing the fastSyncMgr.containsExact() twice? I think the stream() already checked the logic.

Reply:
The second call is applied to different values. The wrapper contains a key and its corresponding value that were requested, plus a set of referenced key-value pairs (other key-value entries that the value above points to).